src/core/batch.rkt

```diff
@@ -6,203 +6,105 @@
-  (define (munge-include-approx prog)
-    (set! size (+ 1 size))
-    (define node ; This compiles to the register machine
-      (match prog
-        [(approx spec impl) (approx spec (munge-include-approx impl))]
-        [(list op args ...) (cons op (map munge-include-approx args))]
-        [_ prog]))
-    (hash-ref! exprhash
-               node
-               (lambda ()
-                 (begin0 (+ exprc varc) ; store in cache, update exprs, exprc
-                   (set! exprc (+ 1 exprc))
-                   (set! icache (cons node icache))))))
+(struct batch ([nodes #:mutable] [roots #:mutable] vars))
 
-  (define roots
-    (list->vector (map (if ignore-approx munge-ignore-approx munge-include-approx) exprs)))
-  (define nodes (list->vector (reverse icache)))
-  (define nodes-length (vector-length nodes))
+(define (batch-length b)
+  (cond
+    [(batch? b) (vector-length (batch-nodes b))]
+    [(mutable-batch? b) (hash-count (mutable-batch-index b))]
+    [else (error 'batch-length "Invalid batch" b)]))
 
-  (when timeline-push
-    (timeline-push! 'compiler (+ varc size) (+ exprc varc)))
-  (batch nodes roots vars nodes-length))
+(struct mutable-batch ([nodes #:mutable] [index #:mutable] [vars #:mutable]))
 
-(define (batch->progs batch)
-  (define roots (batch-roots batch))
-  (define nodes (batch-nodes batch))
+(define (make-mutable-batch)
+  (mutable-batch '() (make-hash) '()))
 
-  (define (unmunge reg)
-    (define node (vector-ref nodes reg))
-    (match node
-      [(approx spec impl) (approx spec (unmunge impl))]
-      [(list op regs ...) (cons op (map unmunge regs))]
-      [_ node]))
+(define (batch-push! b term)
+  (define hashcons (mutable-batch-index b))
+  (hash-ref! hashcons
+             term
+             (lambda ()
+               (let ([new-idx (hash-count hashcons)])
+                 (hash-set! hashcons term new-idx)
+                 (set-mutable-batch-nodes! b (cons term (mutable-batch-nodes b)))
+                 (when (symbol? term)
+                   (set-mutable-batch-vars! b (cons term (mutable-batch-vars b))))
+                 new-idx))))
 
-  (define exprs
-    (for/list ([root (in-vector roots)])
-      (unmunge root)))
-  exprs)
+(define (mutable-batch->immutable b roots)
+  (batch (list->vector (reverse (mutable-batch-nodes b))) roots (reverse (mutable-batch-vars b))))
 
-(define (expand-taylor input-batch)
-  (define vars (batch-vars input-batch))
-  (define nodes (batch-nodes input-batch))
-  (define roots (batch-roots input-batch))
+(struct batchref (batch idx))
 
-  ; Hash to avoid duplications
-  (define icache (reverse vars))
-  (define exprhash
-    (make-hash (for/list ([var vars]
-                          [i (in-naturals)])
-                 (cons var i))))
-  (define exprc 0)
-  (define varc (length vars))
+(define (deref x)
+  (match-define (batchref b idx) x)
+  (expr-recurse (vector-ref (batch-nodes b) idx) (lambda (ref) (batchref b ref))))
```

Comment:
I thought of doing this as `delay` function in Racket, but maybe this way is better

Reply:
Yeah I worry that would be hard to control